File: alogorithms/leetcode/isValidParentheses.py
def isValid(s):
    stack = list()
    for char in s:
        if char == '(' or char == '{' or char =='[':
            print(char , "push")
            stack.append(char)
        else:
            if char == ')':
                print(char, stack[-1])
                if not stack or stack[-1] != '(':
                    print("here 1")
                    return False
                stack.pop()
            elif char == '}':
                if not stack or stack[-1] != '{':
                    print("here 2")
                    return False
                stack.pop()
            elif char == ']':
                if not stack or stack[-1] != '[':
                    print("here 3")
                    return False
                stack.pop()
    
    if stack:
        print("here 4")
        return False
    else:
        return True

print(isValid("()[]{}"))

File: AL/isValidParentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
  const stack = [];

  for (const c of s) {
    if (c === "(" || c === "{" || c === "[") {
      stack.push(c);
    } else if (c === ")" || c === "}" || c === "]") {
      if (stack.length === 0) return false;
      let open = stack.pop();
      switch (c) {
        case ")":
          if (open !== "(") return false;
          break;
        case "}":
          if (open !== "{") return false;
          break;
        case "]":
          if (open !== "[") return false;
          break;
      }
    }
  }

  return stack.length === 0;
};
